primary/tests/integration_tests.rs

```diff
@@ -0,0 +1,215 @@
+
+    // Create a new test payload store.
+    let payload_map = rocks::DBMap::<(BatchDigest, WorkerId), PayloadToken>::open(
+        temp_dir(),
+        None,
+        Some("payloads"),
+    )
+    .unwrap();
+    let payload_store = Store::new(payload_map);
+
+    // Create a new test batch store.
+    let batch_map = rocks::DBMap::<BatchDigest, SerializedBatchMessage>::open(
+        temp_dir(),
+        None,
+        Some("batches"),
+    )
+    .unwrap();
+    let batch_store = Store::new(batch_map);
+
+    let worker_id = 0;
+    // Headers/Collections
+    let mut header_ids = Vec::new();
+    // Blocks/Certficates
+    let mut block_ids = Vec::new();
+    let key = keys().pop().unwrap();
+    let mut missing_batch = CertificateDigest::new([0; 32]);
+
+    // Generate headers
+    for n in 0..5 {
+        let batch = fixture_batch_with_transactions(10);
+
+        let header = fixture_header_builder()
+            .with_payload_batch(batch.clone(), worker_id)
+            .build(|payload| key.sign(payload));
+
+        let certificate = certificate(&header);
+        let block_id = certificate.digest();
+
+        // Write the certificate
+        certificate_store
+            .write(certificate.digest(), certificate.clone())
+            .await;
+
+        // Write the header
+        header_store.write(header.clone().id, header.clone()).await;
+
+        header_ids.push(header.clone().id);
+
+        // Write the batches to payload store
+        payload_store
+            .write_all(vec![((batch.clone().digest(), worker_id), 0)])
+            .await
+            .expect("couldn't store batches");
+
+        block_ids.push(block_id);
+
+        if n != 4 {
+            // Add a batch to the workers store
+            let message = WorkerMessage::<Ed25519PublicKey>::Batch(batch.clone());
+            let serialized_batch = bincode::serialize(&message).unwrap();
+            batch_store.write(batch.digest(), serialized_batch).await;
+        } else {
+            missing_batch = block_id;
+        }
+    }
+
+    let (tx_new_certificates, _rx_new_certificates) = channel(CHANNEL_CAPACITY);
+    let (_tx_feedback, rx_feedback) = channel(CHANNEL_CAPACITY);
+
+    Primary::spawn(
+        name.clone(),
+        signer,
+        committee.clone(),
+        parameters.clone(),
+        header_store,
+        certificate_store,
+        payload_store,
+        /* tx_consensus */ tx_new_certificates,
+        /* rx_consensus */ rx_feedback,
+    );
+
+    // Wait for primary to start all components (including grpc server)
+    tokio::time::sleep(Duration::from_secs(1)).await;
+
+    // Spawn a `Worker` instance.
+    Worker::spawn(
+        name.clone(),
+        worker_id,
+        committee.clone(),
+        parameters.clone(),
+        batch_store,
+    );
+
+    // Test gRPC server with client call
+    let mut client = ValidatorClient::connect("http://127.0.0.1:50052")
+        .await
+        .unwrap();
```

Comment:
I see that you put a `sleep` above to ensure that you give some time for the primary to spin up but there is always the possibility for this to take more time. Maybe instead we want to retry here in case where the client can not connect successfully? So in this case instead of adding a sleep on line 133, we do a few retries (with some wait before each retry step) and fail permanently when maxed the retries?

Reply:
Ideally we'd bind on port zero, query what port we actually are bound to and then use that port to completely avoid port conflicts as well as sleeps in tests

Reply:
@akichidis just to clarify before I make the change, I would have to add sleeps to do the waits before each retry step correct? Or is there another way to do the wait you are referring to here. 

@bmwill looks like I will need to do this in this PR to get [local benchmarking working](https://github.com/MystenLabs/narwhal/pull/140#pullrequestreview-958415847) as @huitseeker pointed out. If I bind to `127.0.0.1:0` when I spawn the gRPC server a port will dynamically assigned, which is handled by some internal tonic code. How would I be able to query that port from the client? Do I have to add some custom code here or is there an easier way to do this?